### PR TITLE
add num_error visualisation on demand

### PR DESCRIPTION
**Pull request by akrzgc**
The change introduces functionality of numerical error visualisation for selected tensors (all possible at the moment).
In order to visualise numerical error for a given `nanoDFT.py` run one should use new parameter `--vis_num_error true`.
The script uses host callback to be able to save the tensors from the device to a number of files so that option shall not be used when profiling performance.
The script assumes that the parameter provided as a `--mol_str` from the command line will be treated as molecule name and based on that the name of required helper directories and files will be created.
However, one may also use a newly introduced parameter `--molecule_name` which would overwrite the `--mol_str` provided by the CLI.

The script executed with `--vis_num_error true` shall provide similar result to below ones:
visualize_DFT_num_error_benzene
visualize_DFT_num_error_c20

The change is addressing the first part of: https://github.com/graphcore-research/pyscf-ipu/issues/43

### CODE REVIEW

**Comment by AlexanderMath** [comment]
Looks good. Only suggestion: I think if we remove the `+15` [here](https://github.com/graphcore-research/pyscf-ipu/blob/dbf4e92f29be19f65a7e1990a39f72b57c3b12ce/pyscf_ipu/nanoDFT/utils.py#L206C34-L206C34) the label would be put on top of the line.

File: pyscf_ipu/nanoDFT/nanoDFT.py [hunk 1]
```diff
@@ -81,8 +81,27 @@ def nanoDFT_iteration(i, vals, opts, mol):
     log["matrices"] = jax.lax.dynamic_update_slice(log["matrices"], H.             reshape(1, 1, N, N), (i, 2, 0, 0))
     log["energy"]   = log["energy"].at[i].set(energy(density_matrix, H_core, diff_JK, E_xc, E_nuc))  # (iterations, 6)
 
+    if opts.vis_num_error is True:
```
Comment (AlexanderMath) [inline pyscf_ipu/nanoDFT/nanoDFT.py:84]: Neat. I debated with myself whether it'd be nice to have it in separate function. I prefer current form: I like that current form allow quick change of callback which is neat.

File: pyscf_ipu/nanoDFT/nanoDFT.py [hunk 2]
```diff
@@ -562,6 +583,10 @@ def nanoDFT_options(
         ao_threshold (float): Zero out grid_AO that are below the threshold in absolute value.
         dense_ERI (bool): Whether to use dense ERI (s1) or sparse symmtric ERI. 
     """
+    if molecule_name is None:
```
Comment (AlexanderMath) [inline pyscf_ipu/nanoDFT/nanoDFT.py:586]: Ok

File: pyscf_ipu/nanoDFT/nanoDFT.py [hunk 3]
```diff
@@ -603,6 +628,12 @@ def main():
         nanoDFT_forces = grad(mol, grid_coords, grid_weights, mo_coeff, mo_energy)
         pyscf_E, pyscf_hlgap, pyscf_forces = pyscf_reference(mol_str, opts)
         print_difference(nanoDFT_E, nanoDFT_forces, nanoDFT_logged_E, nanoDFT_hlgap, pyscf_E, pyscf_forces, pyscf_hlgap)
+
```
Comment (AlexanderMath) [inline pyscf_ipu/nanoDFT/nanoDFT.py:631]: Ok

File: pyscf_ipu/nanoDFT/utils.py [hunk 4]
```diff
@@ -142,4 +142,80 @@ def min_interatomic_distance(mol_str):
...
+            if i == 0:
+                xticks.append(ys[0])
+                xticklabels.append(file.replace(".npz", "").replace("%i_"%i, ""))
+
+        plt.plot( [10**(-10), 10**(-10)], [0, xticks[-1]], 'C7--', alpha=0.6)
+        plt.plot( [10**(10), 10**10], [0, xticks[-1]], 'C7--', alpha=0.6)
+        plt.plot( [10**(0), 10**0], [0, xticks[-1]], 'C7-', alpha=1)
+
+        for x, label in zip(xticks, xticklabels):
+            ax.text(1e10, x+0.25, label, horizontalalignment='left', size='small', color='black', weight='normal')
```
Comment (AlexanderMath) [inline pyscf_ipu/nanoDFT/utils.py:206]: I think if we remove the 0.25 here the text would go on top of the line which may look at little nicer.

**Review by AlexanderMath — COMMENTED** [review]
Looks good! Only comment: The `+0.25` in label utils.py line 206 may be removed/adjusted to text appear on top of the lines.

**Review by AlexanderMath — APPROVED** [review]
Sorry clicked "comment" instead of "approve". Feel free to decide whether text should be on top of lines by modifying the `+0.25`.

File: pyscf_ipu/nanoDFT/utils.py (hunk 4 above)
Comment (akrzgc) [reply pyscf_ipu/nanoDFT/utils.py:206]: That's true but in case the tensor values go beyond 10e10 the text would cover the dots. We might anticipate that it will never happen but as it is now it is more generic being ready for much higher values.
